src/retention/sqlite/statements.rs

```diff
@@ -0,0 +1,1124 @@
+// This file is part of Astarte.
+//
+// Copyright 2024 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{borrow::Cow, collections::HashSet, time::Duration};
+
+use rusqlite::{Connection, OptionalExtension, Transaction};
+use tracing::{trace, warn};
+
+use crate::{
+    retention::{Id, PublishInfo, StoredInterface},
+    store::sqlite::{
+        statements::{include_query, ReadConnection, WriteConnection},
+        wrap_sync_call, SqliteError,
+    },
+};
+
+use super::{RetentionMapping, RetentionPublish, TimestampSecs};
+
+impl WriteConnection {
+    pub(super) fn store(
+        &mut self,
+        mapping: &RetentionMapping<'_>,
+        publish: &RetentionPublish<'_>,
+    ) -> Result<(), SqliteError> {
+        let exists = read_mapping(self, &mapping.interface, &mapping.path)?.is_some_and(|stored| {
+            if stored != *mapping {
+                warn!("mappings differ, replacing");
+
+                false
+            } else {
+                trace!("mapping already exists");
+
+                true
+            }
+        });
+        let transaction = self.transaction().map_err(SqliteError::Transaction)?;
+
+        wrap_sync_call(|| {
+            if !exists {
+                Self::store_mapping(&transaction, mapping)?;
+                trace!("mapping stored");
+            }
+
+            Self::store_publish(&transaction, publish)?;
+            trace!("publish stored");
+
+            transaction.commit()?;
+
+            Ok(())
+        })
+    }
+
+    pub(super) fn store_mapping(
+        transaction: &Transaction<'_>,
+        mapping: &RetentionMapping<'_>,
+    ) -> Result<(), SqliteError> {
+        let mut statement = transaction
+            .prepare_cached(include_query!("queries/retention/write/store_mapping.sql"))
+            .map_err(SqliteError::Prepare)?;
+
+        let expiry_sec = mapping.expiry_to_sql();
+
+        statement
+            .execute((
+                &mapping.interface,
+                &mapping.path,
+                mapping.version_major,
+                mapping.reliability,
+                expiry_sec,
+            ))
+            .map_err(SqliteError::Query)?;
+
+        Ok(())
+    }
+
+    pub(super) fn store_publish(
+        transaction: &Transaction<'_>,
+        publish: &RetentionPublish<'_>,
+    ) -> Result<(), SqliteError> {
+        let mut statement = transaction
```

Comment:
Why didn't you wrap this part with `wrap_sync_call` as well?

Reply:
Because the store_publish and store_mapping are lower "level" functions for the `store` call, and we can wrap both in a single wrap_sync_call.

This should be private, but I exposed it to be used in the tests.